pczt/src/sapling.rs

```diff
@@ -0,0 +1,526 @@
+    /// - This is required by the Prover, instead of disclosing `shared_secret` to them.
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The value commitment randomness.
+    ///
+    /// - This is set by the Constructor.
+    /// - The IO Finalizer compresses it into the bsk.
+    /// - This is required by the Prover.
+    /// - This may be used by Signers to verify that the value correctly matches `cv`.
+    ///
+    /// This opens `cv` for all participants. For Signers who don't need this information,
+    /// or after proofs / signatures have been applied, this can be redacted.
+    pub(crate) rcv: Option<[u8; 32]>,
+
+    /// The `ock` value used to encrypt `out_ciphertext`.
+    ///
+    /// This enables Signers to verify that `out_ciphertext` is correctly encrypted.
+    ///
+    /// This may be `None` if the Constructor added the output using an OVK policy of
+    /// "None", to make the output unrecoverable from the chain by the sender.
+    pub(crate) ock: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the output.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// Proprietary fields related to the note being spent.
+    #[getset(get_mut)]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
+        // Destructure `other` to ensure we handle everything.
+        let Self {
+            mut spends,
+            mut outputs,
+            value_sum,
+            anchor,
+            bsk,
+        } = other;
+
+        // If `bsk` is set on either bundle, the IO Finalizer has run, which means we
+        // cannot have differing numbers of spends or outputs, and the value balances must
+        // match.
+        match (self.bsk.as_mut(), bsk) {
+            (Some(lhs), Some(rhs)) if lhs != &rhs => return None,
+            (Some(_), _) | (_, Some(_))
+                if self.spends.len() != spends.len()
+                    || self.outputs.len() != outputs.len()
+                    || self.value_sum != value_sum =>
+            {
+                return None
+            }
+            // IO Finalizer has run, and neither bundle has excess spends or outputs.
+            (Some(_), _) | (_, Some(_)) => (),
+            // IO Finalizer has not run on either bundle. If the other bundle has more
+            // spends or outputs than us, move them over; these cannot conflict by
+            // construction.
+            (None, None) => {
+                if spends.len() > self.spends.len() {
+                    // TODO: Update `self.value_balance`.
+                    self.spends.extend(spends.drain(self.spends.len()..));
+                }
+                if outputs.len() > self.outputs.len() {
+                    // TODO: Update `self.value_balance`.
```

Comment:
I think these TODOs need to be resolved, otherwise it leaves an incorrect `self.value_balance`. (I see why it is more complicated than in the Orchard case: it may be that one bundle has longer spends, and the other has longer outputs. Maybe we can just return `None` in that case?)

Reply:
You are correct; good catch. I'll just return `None` in the case we can't handle without parsed bundles.